### PR TITLE
Print the egglog program before the last schedule for multi-pass schedule in `--run-mode egglog`

### PR DESCRIPTION
**Pull request by yihozhang**
Follow up of #670

### CODE REVIEW

File: dag_in_context/src/lib.rs
Comment (oflatt) [inline dag_in_context/src/lib.rs:261]: This seems weird... I would expect `-1` to run all but the last pass, but it looks like it runs all of them.
Perhaps you should make it i64::MAX by default and use `[0 .. scheduler.len() + stop_after_n_passes ]` in the neg case?

File: dag_in_context/src/schedule.rs
Comment (oflatt) [inline dag_in_context/src/schedule.rs:176]: good catch

File: src/util.rs
Comment (oflatt) [inline src/util.rs:808]: It would be nice if this took into account stop_after_n_passes so that we can debug particular eggcc passes
Right now it just takes the last pass

File: src/util.rs
Comment (oflatt) [inline src/util.rs:813]: You could refactor the inner loop of optimize into a function so you can get the egglog file directly instead of calling build_program

**Review by oflatt — CHANGES_REQUESTED** [review]
Some suggested improvments

File: src/util.rs
Comment (yihozhang) [reply src/util.rs:808]: Yeah realized this right after I pushed

File: dag_in_context/src/lib.rs
Comment (yihozhang) [reply dag_in_context/src/lib.rs:261]: I updated the code to reflect this, although it is a little less elegant now to handle the pass before `i64::MAX`

File: src/util.rs
Comment (yihozhang) [reply src/util.rs:818]: hmm, I'm not sure how to refactor here. Maybe leave it to a future PR?

Comment (oflatt) [reply src/util.rs:818]: sure

File: src/util.rs
Comment (oflatt) [inline src/util.rs:822]: You're still using the wrong pass here- didn't you want the stop_after_n_passes pass?

**Review by oflatt — CHANGES_REQUESTED** [review]

File: src/util.rs
Comment (yihozhang) [reply src/util.rs]: good catch, fixed!

File: src/util.rs
Comment (oflatt) [inline src/util.rs:813]: Why are you subtracting 1 here? Shouldn't you pass it to the config as-is?

**Review by oflatt — CHANGES_REQUESTED** [review]

File: src/util.rs
Comment (yihozhang) [reply src/util.rs:813]: say there are N passes and the user passes `stop-after-n-passes N` (the default value). This will run 0..(N-1) passes, and use the last schedule, which is at position `N-1`. Maybe I miss something here?

Comment (oflatt) [reply src/util.rs:813]: Ah okay, I get it now

**Review by oflatt — APPROVED** [review]
Great work! I just ask that you test this PR before you merge, since the logic is surprisingly subtle
Perhaps we should add a small check that running the result of the egglog run mode is the same as running n passes?

**Comment by yihozhang** [comment]
Added a test

**Pull request merged**
